File: basic/01_searching/01_linear_search/LinearSearch.py
##############################################################
# Simple algorithm for linear search in python
# Author : Muhamed Muhsin A. <muhsinibnalazeez@gmail.com>
# 
# Run : python LinearSearch.py
##############################################################

from typing import Union, Sequence

def linear_search(arr: Sequence[Union[int, float]],
                  key: Union[int, float]) -> int:
    """
    Linear Search through Array `arr`
    Input :
        arr :: List of int | float
        key :: Element to search (int | float)
    Output :
        index :: The index of the position where the key 
        occure at first place. returns -1 if not found
    """
    
    # Iterae through the element of arr
    for i in range(len(arr)):
        if arr[i] == key:
            return i
    # key not found
    return -1


#################################################################
if __name__ == "__main__":
    arr_int   = [2, 4, 5, 6, 7, 8, 9, 19, 78, 88, 199, 6]
    arr_float = [4.5, 7.8, 6.8, 9.5, 7.0]
    
    print("Searching through an integer array")
    key: Union[float, int] = 19
    index = linear_search(arr_int, key)
    if index != -1 :
        print(f"Element {key} found at index {index}")
    else :
        print("Element not found")
    
    print()
    print("Searching through a float array")
    key = 1.1
    index = linear_search(arr_float, key)
    if index != -1 :
        print(f"Element {key} found at {index}")
    else :
        print("Element not found")

File: basic/01_searching/01_linear_search/LinearSearch.java
/**********************************************************************
 * Simple linear search in Java
 * Author : Muhammed Muhsin A. <muhsinibnalazeez@gmail.com>
 *
 * Compile : javac LinearSearch.java
 * Run : java LinearSearch
 **********************************************************************/

public class LinearSearch {
  
  static int linearSearch(int[] arr, int key){
    for (int i=0; i< arr.length; i++) {
      if(arr[i] == key) {
        return i;
      }
    }
    return -1;
  }
  static int linearSearch(float[] arr, float key){
    for (int i=0; i<arr.length; i++) {
      if(arr[i] == key) {
        return i;
      }
    }
    return -1;
  }

  public static void main(String[] args){
    int i, key1;
    float key2;
    int[] arr_i   = {6, 55, 34, 11, 99, 67, 4, 34, 12, 3};
    float[] arr_f = {6.5f, 4.3f, 2.0f, 8f, 90.0f, 3.1f, 2.2f, 89.9f, 5.6f, 34.3f};

    System.out.println("searching through an integer array");
    key1 = 34;
    i = linearSearch(arr_i, key1);
    if(i != -1){
        System.out.println("Element " + key1 + " found at index " + i);
    }else{
        System.out.println("Element not found");
    }

    System.out.println("searching through a float array");
    key2 = 3.3f;
    i = linearSearch(arr_f, key2);
    if(i != -1){
        System.out.println("Element " + key2 + " found at index " + i);
    }else{
        System.out.println("Element not found");
    }

 }
}
